### PR TITLE
feat: Forward Booking Questions in Query Parameters #15755

### LINKED ISSUE #15755: Forward Booking Questions in Query Parameters
### Is your proposal related to a problem?

It's frustrating that I'm not able to access the "booking questions" via query parameters on the confirmation page.

(Write your answer here.)

### Describe the solution you'd like

Currently, Cal provides the option to "forward parameters" to the "Redirect on booking" page. However, the issue is that these parameters only include the basic event details.

There is no option/inclusion to forward the "booking questions" inside of the query parameters as well.

We need this data available on the booking page, so it would be great to have the option to forward these data points as well. 

### Describe alternatives you've considered

None, there's no way to solve this since Cal's infrastructure is controlling this entire step.

---
##### House rules
- If this issue has a `🚨 needs approval` label, don't start coding yet. Wait until a core member approves feature request by removing this label, then you can start coding.
  - For clarity: Non-core member issues automatically get the `🚨 needs approval` label.
  - Your feature ideas are invaluable to us! However, they undergo review to ensure alignment with the product's direction.

### PR DESCRIPTION
**Pull request by Flash-Shivam**
## What does this PR do?

- Fixes #15755 (GitHub issue number)

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com). If N/A, write N/A here and check the checkbox.
- [ ] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

Working demo Video : https://drive.google.com/file/d/1FqRMC5Z6Jr8SFO-dIxizKRzopioCVdWm/view?usp=sharing

Output redirected link : `https://github.com/Flash-Shivam?user=shivam&type=test-event&isSuccessBookingPage=true&email=shivamjadhav9333%40gmail.com&eventTypeSlug=test-event&title=Test+Event+between+Nameless+and+shivam&description=Notes&startTime=2024-08-16T08%3A30%3A00.000Z&endTime=2024-08-16T08%3A45%3A00.000Z&location=integrations%3Adaily&responses.name=shivam&responses.Alias=jadshiv&responses.email=shivamjadhav9333%40gmail.com&responses.notes=Notes&responses.guests=shivamjadhav9333%40gmail.com `

### CODE REVIEW

File: packages/lib/bookingSuccessRedirect.ts
Comment (anikdhabal) [inline packages/lib/bookingSuccessRedirect.ts:57]: It works fine, but looks like type error is visible. Pls fix that

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:57]: Do you want me to do this 
```
const responseParams = booking.responses
  ? Object.keys(booking.responses).reduce((obj, key) => {
      if (Array.isArray(booking.responses[key])) {
        obj[`responses.${key}`] = booking.responses[key].join(",");
      } else {
        obj[`responses.${key}`] = booking.responses[key] as string;
      }
      return obj;
    }, {} as Record<string, string>)
  : {};
  ```

Comment (anikdhabal) [reply packages/lib/bookingSuccessRedirect.ts:57]: It's ok, currently booking response is unknown type, as the type "responses" is not included in the union type of the constraints specified.

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:57]: if it looks good can you approve the PR. Also do I need to get it approved someone as well ?

Comment (anikdhabal) [reply packages/lib/bookingSuccessRedirect.ts:57]: No you don't approve if all the checks are not passed. The type error visible globally, pls fix that

**Review by Amit91848 — CHANGES_REQUESTED** [review]
Thanks for the pr @Flash-Shivam !

- Location field needs some more refinement, it does not work properly.
  ?user=teampro&type=quick-chat&isSuccessBookingPage=true&email=s%40s.com&eventTypeSlug=quick-chat&title=Quick+Chat+between+Team+Pro+Example+and+s&description=&startTime=2024-08-16T08%3A00%3A00.000Z&endTime=2024-08-16T08%3A15%3A00.000Z&location=%2B91923404345&responses.name=s&responses.email=s%40s.com&responses.guests=&`responses.location=%5Bobject+Object%5D`
  
 -  For conferencing locations, it should have the meeting link and not integrations:daily etc.
     ?user=shivam&type=test-event&isSuccessBookingPage=true&email=shivamjadhav9333%40gmail.com&eventTypeSlug=test-event&title=Test+Event+between+Nameless+and+shivam&description=Notes&startTime=2024-08-16T08%3A30%3A00.000Z&endTime=2024-08-16T08%3A45%3A00.000Z&`location=integrations%3Adaily`&responses.name=shivam&responses.Alias=jadshiv&responses.email=shivamjadhav9333%40gmail.com&responses.notes=Notes&responses.guests=shivamjadhav9333%40gmail.com

**Comment by Flash-Shivam** [comment]
> Thanks for the pr @Flash-Shivam !
> 
> * Location field needs some more refinement, it does not work properly.
>   ?user=teampro&type=quick-chat&isSuccessBookingPage=true&email=s%40s.com&eventTypeSlug=quick-chat&title=Quick+Chat+between+Team+Pro+Example+and+s&description=&startTime=2024-08-16T08%3A00%3A00.000Z&endTime=2024-08-16T08%3A15%3A00.000Z&location=%2B91923404345&responses.name=s&responses.email=s%40s.com&responses.guests=&`responses.location=%5Bobject+Object%5D`
> * For conferencing locations, it should have the meeting link and not integrations:daily etc.
>   ?user=shivam&type=test-event&isSuccessBookingPage=true&email=shivamjadhav9333%40gmail.com&eventTypeSlug=test-event&title=Test+Event+between+Nameless+and+shivam&description=Notes&startTime=2024-08-16T08%3A30%3A00.000Z&endTime=2024-08-16T08%3A45%3A00.000Z&`location=integrations%3Adaily`&responses.name=shivam&responses.Alias=jadshiv&responses.email=shivamjadhav9333%40gmail.com&responses.notes=Notes&responses.guests=shivamjadhav9333%40gmail.com

For the location part it is the existing code that needs to be changed , this we get from the api response of `https://cal.com/api/book/event`. Can be taken as part of another separate issue, as this will require approval from cal.com team member. 

I have modified the code to  take care of the Location field in responses.

**Comment by anikdhabal** [comment]
> For the location part it is the existing code that needs to be changed , this we get from the api response of `https://cal.com/api/book/event`. Can be taken as part of another separate issue, as this will require approval from cal.com team member.
> 
> I have modified the code to take care of the Location field in responses.

Don't panic. In a PR, we test all cases and checks before approving and merging, so everything should be reviewed. The PR is still incomplete and needs changes. Also, which issue are you referring to? If possible, try to address it. The PR should remain incomplete if that issue is not fixed

File: packages/lib/bookingSuccessRedirect.ts
Comment (anikdhabal) [inline packages/lib/bookingSuccessRedirect.ts:50]: "There is still a problem with the location response. You don't get videocallurl for the dynamic video app in the location, so you’re seeing something like `integration_.....`

**Review by anikdhabal — CHANGES_REQUESTED** [review]

File: packages/lib/bookingSuccessRedirect.ts
Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:50]: Can you paste the output redirected link and the response of this api : `https://cal.com/api/book/event`

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:50]: Exactly what I said in the earlier comment, currently the params that are passed are taken from have responses object (from the api call : `https://cal.com/api/book/event` )as below
```
"responses": {
    "name": "Magnus",
    "Alias": "jadshiv",
    "email": "shivamjadhav9333@gmail.com",
    "notes": "Notes typos",
    "title": "Chess",
    "guests": [
      "shivam.jadhav1170378@gmail.com"
    ],
    "location": {
      "value": "integrations:google:meet",
      "optionValue": ""
    }
  }
```

So I am using this as it is , if we want to have the video call url we need to change the existing code, who will give the approval for this ?

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:50]: I have added meeting url in responses params.

File: packages/lib/bookingSuccessRedirect.ts
Comment (anikdhabal) [inline packages/lib/bookingSuccessRedirect.ts:68]: Good find but i think instead adding an extra param, overwrite the location when there is meetingLink present. Means location response value would be the meetingurl

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts]: Updatedd

Comment (anikdhabal) [reply packages/lib/bookingSuccessRedirect.ts]: Pls fix the type error now

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts]: Which typo error check the latest commit please

Comment (anikdhabal) [reply packages/lib/bookingSuccessRedirect.ts]: Have you checked the test to see if it has passed? Please check below

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts]: Done

File: packages/lib/bookingSuccessRedirect.ts
Comment (anikdhabal) [inline packages/lib/bookingSuccessRedirect.ts:29]: No this is not correct. Don't set any as type, pls find the actual issue

Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:29]: Currently the code is written in such way that this has to be done. If you know any other way to fix it go ahead.

File: packages/lib/bookingSuccessRedirect.ts
Comment (anikdhabal) [inline packages/lib/bookingSuccessRedirect.ts:69]: We are passing 'location' as the default in the parameters, so you need to filter it out from the response. Otherwise, it will display both 'location:' and 'response.location:'. If 'location' is present, overwrite the video call URL with it

**Review by anikdhabal — COMMENTED** [review]
There could be an additional checkbox to enable or disable the passing of booking responses:-
Screenshot 2024-08-17 085926

File: packages/lib/bookingSuccessRedirect.ts
Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:69]: What's the issue if it shows both `location` and `response.location` because response is something we are getting from the API response.

**Comment by Flash-Shivam** [comment]
> There could be an additional checkbox to enable or disable the passing of booking responses:- Screenshot 2024-08-17 085926

Open a separate issue for this. First let this PR get merged. This is out of the scope of this PR.

File: packages/lib/bookingSuccessRedirect.ts
Comment (Flash-Shivam) [reply packages/lib/bookingSuccessRedirect.ts:69]: Also can we get the PR approved and merged, I think it's functional and we are drifting off for a while now.

**Comment by anikdhabal** [comment]
Hey @Flash-Shivam Is it possible to address all the reviews and complete this feature here from your end? Or let another contributor work on finishing the entire workflow in a PR. This PR is currently incomplete, and we do not merge incomplete PRs.

**Comment by Flash-Shivam** [comment]
> Hey @Flash-Shivam Is it possible to address all the reviews and complete this feature here from your end? Or let another contributor work on finishing the entire workflow in a PR. This PR is currently incomplete, and we do not merge incomplete PRs.

The feature that you are asking about is redundant as the params will be used to show some data on the custom redirected link, even if the user doesn't want to use it won't harm anyone if they are present. Which reviews are we talking about ?

**Comment by anikdhabal** [comment]
Things to address:
1. Fix the type error (avoid using any).
2. Ensure there is only one location parameter in the URL.
3. Add a checkbox to include booking responses in the success URL.

**Comment by Flash-Shivam** [comment]
> Things to address:
> 
> 1. Fix the type error (avoid using any).
> 2. Ensure there is only one location parameter in the URL.
> 3. Add a checkbox to include booking responses in the success URL.

1. There are no type errors . See that `yarn type-check:ci` runs successfully, there is reason why `any` is used without which this can't be done, if you can go ahead. 
2. There is only one location parameter in the URL the other one is responses.location which gives metingURL or the responses.location value from the API response . This is what expected check the [original issue](https://github.com/calcom/cal.com/issues/15755#issuecomment-2230637096). 
3. This needs design approval and is top on which can be done later. 


@ciaranha @hariombalhara please take a call on this

**Comment by Flash-Shivam** [comment]
@anikdhabal Can you open this PR I want to work on the remaining feature ?

**Comment by Amit91848** [comment]
Closing because of staleness

**Pull request closed without merging**
